File: Backend/services/blog/service/BlogService.go
package service

import (
	"fmt"

	"blog.xws.com/model"
	"blog.xws.com/repository"
	"github.com/google/uuid"
)

type BlogService struct {
	BlogRepository *repository.BlogRepository
}

func (service *BlogService) GetAllBlogs() ([]model.Blog, error) {
	return service.BlogRepository.GetAll()
}

func (service *BlogService) GetById(id uuid.UUID) (*model.Blog, error) {
	blog, err := service.BlogRepository.GetById(id)
	if err != nil {
		return nil, fmt.Errorf(fmt.Sprintf("Blog with id %s not found", id))
	}

	return &blog, nil
}

func (service *BlogService) Create(blog *model.Blog) error {
	err := service.BlogRepository.Create(model.BeforeCreate(blog.UserId, blog.Title, blog.Description, blog.Images))
	if err != nil {
		return err
	}
	return nil
}

func (service *BlogService) Delete(id uuid.UUID) error {
	return service.BlogRepository.Delete(id)
}

func (service *BlogService) Update(id uuid.UUID, updatedBlog model.Blog) error {
	return service.BlogRepository.Update(id, updatedBlog)
}

func (service *BlogService) GetAllByUser(userId string) ([]model.Blog, error) {
	return service.BlogRepository.GetAllByUser(userId)
}

File: Frontend/frontend-app/src/app/blog/blog.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Blog } from './blog.model';
import { BlogDto } from './blog.dto';

@Injectable({
  providedIn: 'root',
})
export class BlogService {
  private apiUrl = 'http://localhost:8080/blogs';

  constructor(private http: HttpClient) {}

  getAllBlogs(): Observable<Blog[]> {
    return this.http.get<Blog[]>(`${this.apiUrl}`);
  }

  getById(id: string): Observable<Blog> {
    return this.http.get<Blog>(`${this.apiUrl}/${id}`);
  }

  create(blog: BlogDto, imageFiles?: File[]): Observable<any> {
    const formData = new FormData();
    formData.append('userId', blog.userId);
    formData.append('title', blog.title);
    formData.append('description', blog.description);

    console.log('Creating blog with:', {
      userId: blog.userId,
      title: blog.title,
      description: blog.description,
      imageCount: imageFiles?.length || 0,
    });

    if (imageFiles && imageFiles.length > 0) {
      imageFiles.forEach((file, index) => {
        console.log(
          `Appending image ${index}: ${file.name}, size: ${file.size}, type: ${file.type}`
        );
        formData.append('images', file);
      });
    }

    return this.http.post(`${this.apiUrl}`, formData);
  }

  update(id: string, blog: Blog): Observable<any> {
    return this.http.put(`${this.apiUrl}/${id}`, blog);
  }

  delete(id: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }

  getAllByUser(userId: string): Observable<Blog[]> {
    return this.http.get<Blog[]>(`${this.apiUrl}/users/${userId}`);
  }
}
